crates/ruff_dev/src/check_formatter_stability.rs

```diff
@@ -84,8 +76,88 @@ pub(crate) fn main(args: &Args) -> anyhow::Result<ExitCode> {
     }
 }
 
+enum Message {
+    Start {
+        path: PathBuf,
+    },
+    Failed {
+        path: PathBuf,
+        error: anyhow::Error,
+    },
+    Finished {
+        path: PathBuf,
+        result: CheckRepoResult,
+    },
+}
+
+fn check_multi_project(args: &Args) -> bool {
+    let mut all_success = true;
+    let mut total_errors = 0;
+    let mut total_files = 0;
+    let start = Instant::now();
+
+    rayon::scope(|scope| {
+        let (sender, receiver) = channel();
+
+        for base_dir in &args.files {
+            for dir in base_dir.read_dir().unwrap() {
+                let path = dir.unwrap().path().clone();
+                sender.send(Message::Start { path: path.clone() }).unwrap();
+
+                let sender = sender.clone();
+
+                scope.spawn(move |_| {
+                    match check_repo(&Args {
+                        files: vec![path.clone()],
+                        ..*args
+                    }) {
+                        Ok(result) => sender.send(Message::Finished { result, path }),
+                        Err(error) => sender.send(Message::Failed { error, path }),
+                    }
+                    .unwrap();
+                });
+            }
+        }
+
+        scope.spawn(|_| {
+            let mut stdout = stdout().lock();
+
+            for message in receiver {
```

Comment:
thread pools and generators would be a dream combination for writing this kind of code, basically you would write a `par_iter` and flatten over the messages